staging/src/k8s.io/apiserver/pkg/admission/metrics.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"k8s.io/api/admissionregistration/v1alpha1"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	namespace = "apiserver"
+	subsystem = "admission"
+)
+
+var (
+	latencyBuckets       = prometheus.ExponentialBuckets(10000, 2.0, 8)
+	latencySummaryMaxAge = 5 * time.Hour
+
+	Metrics = newAdmissionMetrics()
+)
+
+type AdmissionMetrics struct {
+	step            *metricSet
+	controller      *metricSet
+	externalWebhook *metricSet
+}
+
+func newAdmissionMetrics() *AdmissionMetrics {
+	// Admission metrics for a step of the admission flow. The entire admission flow is broken down into a series of steps
+	// Each step is identified by a distinct type label value.
+	step := newMetricSet("step_",
+		[]string{"operation", "group", "version", "resource", "subresource", "type", "is_system_ns", "rejected"},
+		"Admission sub-step %s, broken out for each operation and API resource and step type (validating or mutating).")
+
+	// Built-in admission controller metrics. Each admission controller is identified by name.
+	controller := newMetricSet("controller_",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource", "is_system_ns", "rejected"},
+		"Admission controller %s, identified by name and broken out for each operation and API resource and type (validating or mutating).")
+
+	// External admission webhook metrics. Each webhook is identified by name.
+	externalWebhook := newMetricSet("external_webhook_",
+		[]string{"name", "type", "operation", "group", "version", "resource", "subresource", "is_system_ns", "rejected"},
+		"External admission webhook %s, identified by name and broken out for each operation and API resource and type (validating or mutating).")
+
+	step.mustRegister()
+	controller.mustRegister()
+	externalWebhook.mustRegister()
+	return &AdmissionMetrics{step: step, controller: controller, externalWebhook: externalWebhook}
+}
+
+func (m *AdmissionMetrics) reset() {
+	m.step.reset()
+	m.controller.reset()
+	m.externalWebhook.reset()
+}
+
+// namedHandler requires each admission.Interface be named, primarly for metrics tracking purposes.
+type NamedHandler interface {
```

Comment:
Comment mismatch: `namedHandler` vs `NamedHandler`

Also IMO it's better to define types before methods

Reply:
You can't have *both* all types before all methods *and* all constructors next to their types (as you request in another comment). ;)

(In this case I agree that NamedHandler ought to go near the top of this file.)

Reply:
Thanks, I've reorganized the types and updated the docs.

Reply:
Sorry, to clarify, I meant it's better to define types not in-between other type's methods, not that all types should go before all methods